polar-core/src/partial/constraints.rs

```diff
@@ -802,7 +816,55 @@ mod test {
         let p = Polar::new();
         p.load_str("f(x) if not print(x);")?;
         let mut q = p.new_query_from_term(term!(call!("f", [partial!("a")])), false);
-        assert!(matches!(q.next_event()?, QueryEvent::Done { .. }));
+        let next = q.next_event()?;
+        assert!(matches!(next, QueryEvent::Done { .. }), "Got {:?}.", next);
+        Ok(())
+    }
+
+    #[test]
+    /// Test an inverted query where a constraint is added inside the inversion that
+    /// is incompatible with existing constraints, but is compatible before it's inverted.
+    fn test_compatible_before_invert() -> TestResult {
+        let p = Polar::new();
+        p.load_str("f(x) if x matches User{} and not (x matches NotUser{} or x.foo = 3);")?;
+
+        let mut query = p.new_query_from_term(term!(call!("f", [partial!("a")])), false);
+        assert_partial_expression!(next_binding(&mut query)?, "a", "_this matches User{} and _this.foo != 3");
+
+        p.load_str("g(x) if x matches User{} and not (x matches NotUser{} and x.foo = 3);")?;
+        let mut query = p.new_query_from_term(term!(call!("g", [partial!("a")])), false);
+        // This output only has _this matches User{}, because we immediately judge x matches NotUser{} to be
+        // incompatible (so the not fails before reaching the and).
+        assert_partial_expression!(next_binding(&mut query)?, "a", "_this matches User{}");
+
+        p.load_str("h(x) if x matches User{} and not (x.foo = 3 and x matches NotUser{});")?;
+        let mut query = p.new_query_from_term(term!(call!("h", [partial!("a")])), false);
+        // This output only has _this matches User{}, because we immediately judge x matches NotUser{} to be
+        // incompatible (so the not fails before reaching the and).
+        assert_partial_expression!(next_binding(&mut query)?, "a", "_this matches User{}");
+
+        Ok(())
+    }
+
+    #[test]
+    #[ignore]
+    // TODO (dhatch): remove ignore after IN can be inverted.
```

Comment:
Is this comment still applicable? I don't see any `in`s in the below test, and we do have tests that ensure we can invert `in`s.